core/src/test/java/org/infinispan/tx/locking/RollbackDuringLockAcquisitionTest.java

```diff
@@ -0,0 +1,188 @@
+             .useSynchronization(useSynchronization)
+             .recovery().disable();
+      builder.locking()
+             .lockAcquisitionTimeout(TimeUnit.MINUTES.toMillis(1));
+
+      cacheManager.defineConfiguration(cacheName, builder.build());
+      Cache<String, String> cache = cacheManager.getCache(cacheName);
+      LockManager lockManager = extractLockManager(cache);
+      TransactionManager tm = cache.getAdvancedCache().getTransactionManager();
+
+      // simulates lock acquired by other transaction
+      lockManager.lock(key, "_tx_", 1, TimeUnit.SECONDS).lock();
+      assertLocked(cache, key);
+
+      tm.begin();
+      CompletableFuture<?> put = cache.putAsync(key, "value1");
+
+      GlobalTransaction gtx = extractComponent(cache, TransactionTable.class).getGlobalTransaction(tm.getTransaction());
+
+      // wait until the tx is queued in the InfinispanLock
+      eventually(() -> lockManager.getLock(key).containsLockOwner(gtx));
+
+      //abort the transaction, simulates the TM's reaper aborting a long running transaction
+      tm.rollback();
+
+      // release the lock. the "put" must not acquire the lock
+      lockManager.unlock(key, "_tx_");
+
+      Exceptions.expectCompletionException(InvalidTransactionException.class, put);
+
+      assertNotLocked(cache, key);
+      assertNoTransactions(cache);
+   }
+
+   @Test(dataProvider = "tm-lookup")
+   public void testRollbackWhileWaitingForLock2(boolean useSynchronization, TransactionManagerLookup lookup)
+         throws Exception {
+      final String cacheName = concat("local-lock", lookup.getClass().getSimpleName(), useSynchronization);
+      final String key = concat("reaper-lock", lookup.getClass().getSimpleName(), useSynchronization);
+
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.transaction()
+             .transactionManagerLookup(lookup)
+             .transactionMode(TransactionMode.TRANSACTIONAL)
+             .lockingMode(LockingMode.PESSIMISTIC)
+             .useSynchronization(useSynchronization)
+             .recovery().disable();
+      builder.locking()
+             .lockAcquisitionTimeout(TimeUnit.MINUTES.toMillis(1));
+
+      cacheManager.defineConfiguration(cacheName, builder.build());
+      Cache<String, String> cache = cacheManager.getCache(cacheName);
+      LockManager lockManager = extractLockManager(cache);
+      TransactionManager tm = cache.getAdvancedCache().getTransactionManager();
+
+      // simulates lock acquired by other transaction
+      lockManager.lock(key, "_tx_", 1, TimeUnit.SECONDS).lock();
+      assertLocked(cache, key);
+
+      AtomicReference<Transaction> tx = new AtomicReference<>();
+
+      Future<Boolean> lock = fork(() -> {
+         tm.begin();
+         tx.set(tm.getTransaction());
+         try {
+            return cache.getAdvancedCache().lock(key);
+         } finally {
+            // we want the exception from lock(key) and not from tm.rollback()
```

Comment:
If `rollback()` if expected to throw an exception, I'd rather the comment said explicitly "Ignore the <AlreadyRolledBackException> from rollback".
If not, then I'd argue it's ok to see the exception from rollback so we can eliminate it.

Reply:
updated the comment with an explanation :)